### DISCUSSION TITLE
Allow to explicitly specify value castings

### DISCUSSION THREAD

**Discussion by andrew-sol** [mikro-orm/mikro-orm#3472] [Q&A] [answered]
Context: PostgreSQL

The problem is that the ORM casts `bigint` and `numeric` types to string in JS and we have no control over it even when we are sure that the value won't exceed the max int range in JS.
Creating custom types/hydrators looks like too much work for such a trivial task.

Let's compare some ranges:
```js
2147483647          // Postgres integer
9007199254740991    // JS max safe integer
9223372036854775807 // Postgres bigint
```

Sure, if the stored value might be greater than "JS max safe integer", we should use the default behaviour ("string" type).
But we can see that the difference between "Postgres integer" and "JS max safe integer" ranges is significant, and often we know for sure that the stored value will never exceed that range. For such cases it would be super-convinient to tell the ORM to always cast the value to number.

Possible API:
```ts
@Property({ type: types.bigint, cast: 'number' })
value: number;
```

**Comment by B4nan** [comment, marked as answer]
The ORM is not casting numbers to strings, it is the database connector (e.g. here the `pg` module) returning them as string.

You _can_ map it to number, as well as you can adjust the mapping on the `pg` level (you can tell `pg` to map it differently via `pg-types` module). The solution is to use a different mapped type than `t.bigint`, a custom one most probably. I dont think we need yet another option, used for this single use case. 

There is quite a lot of issues about this, I am sure you can find some examples.

**Comment by andrew-sol** [comment]
Thanks for the explanation. I thought the ORM does some castings during the mapping process.

Guess I'll have to create a custom type for this purpose.
